### ISSUE TITLE
Swift (data) types?

### ISSUE THREAD

**Issue by AF-cgi** [intuit/GroupedArray#1]
Hey,

nice tool. But what is with the swift data types like AnyObject? I replace it with NSObject. All my object-classes needs the NSObject subclass extension. Is there any way for AnyObject type?

**Comment by smileyborg** [comment]
Hi @AFcgi,

I'm not sure I understand your question. Are you referring to the Objective-C grouped array or the Swift wrapper? Can you share an example of what you're asking?

**Comment by AF-cgi** [comment]
Hey,

I use the Swift wrapper. First, I import the "INTUGroupedArrayImports.h" file. Second, I get the Swift wrapper for Swift 1.2.
My sections are strings, my objects are classes like

```
public class DemoModel {

    public let identifier:String

    public init(identifier:String) {
        self.identifier = identifier
    }
}
```

I create a `MutableGroupedArray<String, AnyObject>`. The error message is:

> String is not conformed with the AnyObject protocol.

I solve this to replace String with NSString, but it's not very well. At least I get a nil object from the groupedArray datasource for _objectAtIndexPath_.
If I reimplement my model class with

```
public class DemoModel: NSObject {
```

and `MutableGroupedArray<NSString, NSObject>` then I get a non-nil object. But in this case I have many other issues. Is there any purely implementation for `GroupedArray<String, AnyObject>`?

**Comment by smileyborg** [comment]
Hi @AFcgi, sorry for the late reply.

Anyways, because the INTUGroupedArray data structure is implemented in Objective-C, there are some restrictions/limitations on the interoperability with Swift, even when you are using the Swift grouped array wrapper.

If you look at the declaration of the `GroupedArray` type [in the Swift code](https://github.com/intuit/GroupedArray/blob/master/Source/Swift/GroupedArray.swift#L38):

```
public class GroupedArray<S: AnyObject, O: AnyObject>
```

you'll see that both S and O (the generic types for the sections and objects, respectively) must be of type `AnyObject`. This restriction is in place because this is how the Objective-C code gets bridged into Swift. The type `AnyObject` can be an instance of any class, but cannot be Swift `struct` or `enum` types.

So the reason you cannot use a Swift `String` type with the grouped array is because `String` is a `struct`. Thankfully, `String` is bridged seamlessly with `NSString` (which is an Objective-C class type), so you can use that instead.

For the case of your custom class, the following code works just fine:

``` swift
class DemoModel {
    let identifier:String

    init(identifier:String) {
        self.identifier = identifier
    }
}

let mutableGroupedArray = MutableGroupedArray<NSString, DemoModel>()
let model = DemoModel(identifier: "hello")
mutableGroupedArray.addObject(model, toSection: "Section 1")
let a = mutableGroupedArray[0, 0]
print("\(a.identifier)")
```

In other words, you can have a grouped array with generics `<NSString, DemoModel>` and this works fine -- neither `a` nor `a.identifier` is nil.

Does this make sense?

**Comment by smileyborg** [comment]
By the way, you may find this article helpful: http://blog.scottlogic.com/2014/09/24/swift-anyobject.html

**Comment by AF-cgi** [comment]
Hi @smileyborg,

I understand your explanation. In my use case I put another model on the same grouped array. The new model class is similar to DemoModel, but it isn't the same. It calls FutureModel.

<code>
let mutableGroupedArray = MutableGroupedArray<NSString, AnyObject>()

let demModel = DemoModel(identifier: "hello")
mutableGroupedArray.addObject(demModel, toSection: "Section 1")

let futureModel = FutureModel(identifier: "world")
mutableGroupedArray.addObject(futureModel, toSection: "Section 1")

let a = mutableGroupedArray[0, 0]

</code>

For this case, a is nil. That is my problem.

**Comment by AF-cgi** [comment]
Hi @smileyborg,

I solve the problem. The example done with Swift 2.0. So I implement a little workaround for my Swift 1.2 project. Thanks for your communication.
